src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,129 @@
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                /// Check if the extension is a key usage extension.
+                /// No point going over it if we have already established that out cert has digital signature
+                if (!foundKeyUsages)
+                {
+                    X509KeyUsageExtension usageExt = extension as X509KeyUsageExtension;
+                    if (usageExt != null)
+                    {
+                        foundKeyUsages = true;
+                        isDigitalSignature = (usageExt.KeyUsages & X509KeyUsageFlags.DigitalSignature) != 0;
+                    }
+                }
+
+                if (foundKeyUsages && foundEku)
+                {
+                    break;
+                }
+            }
+
+            return isClientAuth && isDigitalSignature;
+        }
+
+        /// <summary>
+        ///   Returns a new collection containing valid client certificates from the given X509Certificate2Collection
+        /// </summary>
+        internal static bool TryFindClientCertificate(this X509Certificate2Collection certificates,
+                                                      ISet<string> allowedIssuers,
+                                                      out X509Certificate2 clientCertificate,
+                                                      out X509Chain clientCertChain)
+        {
+            clientCertificate = null;
+            clientCertChain = null;
+            if (certificates == null)
+            {
+                return false;
+            }
+
+            DateTime now = DateTime.Now;
+            foreach(X509Certificate2 cert in certificates)
+            {
+                if (cert.HasPrivateKey && now > cert.NotBefore && now < cert.NotAfter)
+                {
+                    if (IsClientCertificate(cert))
+                    {
+                        X509Chain chain = new X509Chain();
+                        //TODO consider changing the following to AllFlags
+                        chain.ChainPolicy.VerificationFlags = X509VerificationFlags.NoFlag;
+                        chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+                        chain.ChainPolicy.ApplicationPolicy.Add(new Oid(ClientCertificateOid));
+
+                        // We will just build the certificate chain
+                        chain.Build(cert);
+
+                        if (chain.ChainElements.Count > 0)
+                        {
+                            if (allowedIssuers.Count == 0)
+                            {
+                                clientCertificate = cert;
+                                clientCertChain = chain;
+                                return true;
+                            }
+
+                            X509Certificate2 trustAnchor = chain.ChainElements[chain.ChainElements.Count - 1].Certificate;
```

Comment:
This check is only valid if the chain built fully.  You'd know it didn't if chain.ChainStatus.Any(cs => cs.Status == X509ChainStatusFlags.PartialChain).

Note that (at least on Windows, where I just tested this, so hopefully on Unix) setting X509VerificationFlags.AllowUnknownCertificateAuthority will still result in chain.Build to return true if PartialChain is reported.  So you'll really want to check the status flags directly.